flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java

```diff
@@ -0,0 +1,263 @@
+import org.apache.flink.runtime.mailbox.MailboxExecutor;
+import org.apache.flink.runtime.state.CheckpointStorageLocationReference;
+import org.apache.flink.runtime.state.CheckpointStorageWorkerView;
+import org.apache.flink.runtime.state.CheckpointStreamFactory;
+import org.apache.flink.runtime.state.KeyedStateHandle;
+import org.apache.flink.runtime.state.SnapshotResult;
+import org.apache.flink.runtime.state.Snapshotable;
+import org.apache.flink.runtime.state.StateObject;
+import org.apache.flink.runtime.state.changelog.ChangelogStateHandle;
+import org.apache.flink.runtime.state.changelog.SequenceNumber;
+import org.apache.flink.runtime.state.changelog.StateChangelogWriter;
+import org.apache.flink.util.concurrent.ExecutorThreadFactory;
+import org.apache.flink.util.concurrent.FutureUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.RunnableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static java.util.Collections.emptyList;
+import static java.util.Collections.singletonList;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/** Periodically make materialization for the delegated state backend. */
+public class PeriodicMaterializer {
+    private static final Logger LOG = LoggerFactory.getLogger(PeriodicMaterializer.class);
+
+    /**
+     * ChangelogStateBackend only supports CheckpointType.CHECKPOINT; The rest of information in
+     * CheckpointOptions is not used in Snapshotable#snapshot(). More details in FLINK-23441.
+     */
+    private static final CheckpointOptions CHECKPOINT_OPTIONS =
+            new CheckpointOptions(
+                    CheckpointType.CHECKPOINT, CheckpointStorageLocationReference.getDefault());
+
+    /** task mailbox executor, execute from Task Thread. */
+    private final MailboxExecutor mailboxExecutor;
+
+    /** Async thread pool, to complete async phase of materialization. */
+    private final ExecutorService asyncOperationsThreadPool;
+
+    /** scheduled executor, periodically trigger materialization. */
+    private final ScheduledExecutorService periodicExecutor;
+
+    private final CheckpointStreamFactory streamFactory;
+
+    private final Snapshotable keyedStateBackend;
+
+    private final StateChangelogWriter<ChangelogStateHandle> stateChangelogWriter;
+
+    private final AsyncExceptionHandler asyncExceptionHandler;
+
+    private final int allowedNumberOfFailures;
+
+    /** Materialization failure retries. */
+    private final AtomicInteger retries;
+
+    /** Making sure only one materialization on going at a time. */
+    private final AtomicBoolean materializationOnGoing;
```

Comment:
I think this can be boolean (non-atomic) - as it's only accessed only a single (task) thread.

Reply:
yes, I was thinking back-and-force whether this flag to only be touched in the task thread. For now, it can be non-atomic.

Reply:
Let's keep it for now (I think it may also be changed later, if we want to trigger on completion or use pause). Let me refine this after making the final decision.